### DISCUSSION TITLE
How many VCOMs maximum on an 8-endpoint device like the i.mxRT1020?

### DISCUSSION THREAD

**Discussion by ned-pcs** [hathach/tinyusb#1443] [Q&A]
I would like to use TinyUSB to make a multiple USB serial converter out of my MIMXRT1021CAG4A microcontroller. This MCU has eight bidirectional endpoints. After going through the CDC class specification, I still can't figure out how many serial ports I can get out of this. It looks like I would need at least one CDC class interface (at least one control endpoint), but it's unclear whether I could get away with using that one control endpoint as the CDC class interface for each of the other endpoints used as IN/OUT bulk data interfaces.

I read NXP's app note [AN12597](https://www.nxp.com/docs/en/application-note/AN12597.pdf), and looked at their [companion software](https://www.nxp.com/docs/en/application-note-software/AN12597SW.zip). In the app note they say this:

> A CDC class device consists of two subclass interfaces: a CDC class interface and a data class interface. A CDC device can contain zero or more data interfaces.
> • The CDC class interface uses a standard interface descriptor that requires a control endpoint and an optional interrupt IN endpoint.
> • The data class interface requires a Bulk IN and a Bulk OUT type of endpoint. But since the 16 endpoints on the K32L2B3/ K32L2A are bidirectional endpoints, an endpoint can implement both input and output functions, ie the functions of Bulk IN and Bulk OUT can be implemented by one endpoint.
> In order to support more VCOM, the interrupt IN endpoint in CDC class interface can be removed, so that in addition to a common control endpoint only one bidirectional endpoint is needed to implement a VCOM. That is, except that EP0 is used as the control endpoint, the other 15 endpoints can implement 15 VCOMs and this application note implement such a feature. The descriptor structure of the composite class containing 15 CDC subclasses used in this application note is shown in Figure 1.

This description, however, conflicts with the diagram that follows, in which each interface descriptor has a CDC class device with a single control endpoint.

The [discussion on the NXP forum](https://community.nxp.com/t5/Kinetis-Microcontrollers/AN12597-USB-to-Multi-VCOM-on-K32L2-Series-MCU-KL27-on-Win-7/m-p/1154342), though, indicates that the 15-port solution using a 16-bidirectional-endpoint device actually works under Windows 10. (I need to get this to work eventually under Windows 7 Embedded, but that's a different problem).

There is also the uTasker demo, whose operation on the KL27 is [discussed](https://community.nxp.com/t5/Kinetis-Microcontrollers/MKL27Z64-USB-EP1-IN-OUT-data-tansfer/m-p/1073557) in the NXP forum.

I'm unclear as to how to make this work with TinyUSB, though. Is getting six or seven virtual COM ports out of an 8-endpoint MCU like the i.mxRT1024 possible?

**Comment by hathach** [comment]
- iMXRT has 8 bi-direction = 1 Control EP + 7 IN + 7 Out.
- Each cdc requires 2 IN ep + 1 OUT ep (all non-control)

--> max cdc interfaces would be 3.

**Reply by ned-pcs** [reply]
From the [discussion](https://community.nxp.com/t5/Kinetis-Microcontrollers/MKL27Z64-USB-EP1-IN-OUT-data-tansfer/m-p/1073557) on the NXP forum and from their [app note](https://www.nxp.com/docs/en/application-note/AN12597.pdf), it appears that the interrupt IN endpoint is not required (assuming that you don't want to control data rate or modem control signals from the host).

In that case, I guess that for each VCOM, there would be an IAD with a CDC class interface with 0 endpoints and a data class interface with two bulk endpoints (IN and OUT).

How would I do this using TinyUSB?

**Reply by hathach** [reply]
per CDC specs, notification is indeed optional, though I had some issue with windows/linux driver when skipping the notification a long time ago, I forgot the detail, but it is rather old host. Maybe thing has changes now. You could try it now, for implementation, you only need to provide your own configuration descriptor without notification endpoint.

PS: tinyusb in fact does not use notification endpoint so far.
